WKWebView.loadData should not swap BrowsingContextGroups

achristensen07 opened the pull request:
#### 19b2526aa86b5cd8da3d64c55f41030a327e568f
<pre>
WKWebView.loadData should not swap BrowsingContextGroups
<a href="https://bugs.webkit.org/show_bug.cgi?id=292785">https://bugs.webkit.org/show_bug.cgi?id=292785</a>
<a href="https://rdar.apple.com/151020111">rdar://151020111</a>

Reviewed by Ryosuke Niwa.

289035@main introduced a condition to swap BrowsingContextGroups when navigation.isRequestFromClientOrUserInput()
returns true.  This is needed for most forms of requests from client or user input, but not from WKWebView.loadData
or WKWebView.loadHTMLString (which calls WKWebView.loadData).  Otherwise, with site isolation enabled, window.opener
is lost after loading data from the client, which doesn&apos;t happen with site isolation off.
isRequestFromClientOrUserInput is used for many other things, so we don&apos;t want to change its value, but instead we
introduce a new bit to keep this behavior the same with site isolation on.

This is one of at least two changes needed to get TestWebKitAPI.SOAuthorizationPopUp.InterceptionSucceedCloseByItself
working with site isolation enabled, but it seemed worth making its own PR with its own test.

* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::markAsFromLoadData):
(API::Navigation::isFromLoadData const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadData):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, LoadStringAfterOpen)):

Canonical link: <a href="https://commits.webkit.org/294754@main">https://commits.webkit.org/294754@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e51c93f8dff3a905bb38b6b9d0bca07d4b1ee64b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102970 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22644 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12964 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108138 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53611 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105009 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31145 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78278 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/35228 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105976 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17786 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/92908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58613 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17621 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/10951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52968 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87432 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/11018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110511 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30107 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22176 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87265 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30471 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/89111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/86888 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31739 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/9450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16703 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30034 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/35356 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29842 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33169 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31404 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -204,11 +206,12 @@ class Navigation : public ObjectImpl<Object::Type::Navigation> {
+    bool m_userContentExtensionsEnabled : 1 { true };
+    bool m_isLoadedWithNavigationShared : 1 { false };
+    bool m_requestIsFromClientInput : 1 { false };
+    bool m_fromLoadData : 1 { false };
```
Shouldn't this be named m_isFromLoadData?

achristensen07 committed:
WKWebView.loadData should not swap BrowsingContextGroups
https://bugs.webkit.org/show_bug.cgi?id=292785
rdar://151020111

Reviewed by Ryosuke Niwa.

289035@main introduced a condition to swap BrowsingContextGroups when navigation.isRequestFromClientOrUserInput()
returns true.  This is needed for most forms of requests from client or user input, but not from WKWebView.loadData
or WKWebView.loadHTMLString (which calls WKWebView.loadData).  Otherwise, with site isolation enabled, window.opener
is lost after loading data from the client, which doesn't happen with site isolation off.
isRequestFromClientOrUserInput is used for many other things, so we don't want to change its value, but instead we
introduce a new bit to keep this behavior the same with site isolation on.

This is one of at least two changes needed to get TestWebKitAPI.SOAuthorizationPopUp.InterceptionSucceedCloseByItself
working with site isolation enabled, but it seemed worth making its own PR with its own test.

* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::markAsFromLoadData):
(API::Navigation::isFromLoadData const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadData):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, LoadStringAfterOpen)):

Canonical link: https://commits.webkit.org/294754@main
